File: adapters/java/src/test/java/adapter/UtilTest.java
package adapter;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class UtilTest {
    @Test
    void roundTrips() {
        assertEquals("deadbeef", Util.hexEncode(new byte[] { (byte) 0xde, (byte) 0xad, (byte) 0xbe, (byte) 0xef }));
        assertArrayEquals(new byte[] { (byte) 0xde, (byte) 0xad, (byte) 0xbe, (byte) 0xef }, Util.hexDecode("deadbeef"));
    }

    @Test
    void rejectsOddLength() {
        assertThrows(IllegalArgumentException.class, () -> Util.hexDecode("abc"));
    }

    @Test
    void rejectsNonAsciiWithoutCrashing() {
        assertThrows(IllegalArgumentException.class, () -> Util.hexDecode("café"));
    }
}

File: adapters/kotlin/src/test/kotlin/UtilTest.kt
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class UtilTest {
    @Test
    fun roundTrips() {
        assertEquals("deadbeef", hexEncode(byteArrayOf(0xde.toByte(), 0xad.toByte(), 0xbe.toByte(), 0xef.toByte())))
        assertEquals(
            listOf(0xde, 0xad, 0xbe, 0xef).map { it.toByte() },
            hexDecode("deadbeef").toList(),
        )
    }

    @Test
    fun rejectsOddLength() {
        assertFailsWith<IllegalArgumentException> { hexDecode("abc") }
    }

    @Test
    fun rejectsNonAsciiWithoutCrashing() {
        assertFailsWith<IllegalArgumentException> { hexDecode("café") }
    }
}
